Partial support for metaclasses in Python 2

TRManderson opened the pull request:
Works with `__metaclass__ = M`, does not work with inner class (`class __metaclass__(type):`) declarations.

This feels a bit hacky, but seems to consistently work.

elazarg commented:
Maybe this should happen at the same time it happens for Python3, i.e. inside [`fastparse2.py`](https://github.com/python/mypy/blob/master/mypy/fastparse2.py#L458-L468)?

TRManderson commented:
You know, I think you're right. My first implementation was in `fastparse2.py` but I moved it here in an attempt to maybe get the inline class declaration style.

My implementation turned `'<error>'` in `semanal.py` into `'<error>: dynamic'` and added a `'<error>: inner class'` case.

elazarg commented:
Another approach would be to analyze metaclasses in python2 only after the body was analyzed, This approach has its problems, but will give you the assignment "for free" - maybe for inner classes too,

elazarg commented:
#2823

TRManderson commented:
The issue with doing metaclass analysis after the body is analysed is that using metaclass features in methods will be marked as broken.

I've just reimplemented my `fastparse2.py` version, but it ends up introducing a whole bunch of different error states, which I really don't like (as there's no tangible link to the code that reports them as errors).

The ideal result here would be to improve the metaclass code so that it can deal with more than just symbol table lookups (ie function calls, inner classes).

Function calls are not currently considered to be valid in Python 3 metaclass declarations either, so it'd be nice to have those working.

TRManderson commented:
With the latest 3 commits I think I'm now happy for this PR to be merged (assuming tests pass).

JukkaL commented:
Thanks for implementing this! I'm going to fix a few minor issues in a separate PR.

ambv commented:
@JukkaL, this breaks typeshed CI: https://travis-ci.org/python/typeshed/jobs/203959662

gvanrossum commented:
I wonder if this isn't too strict? Here's some code that fails now:
```python
from typing import Any
metabase = None  # type: Any
class M(metabase): pass
class C(object):
  __metaclass__ = M
```
This complains that
```
__tmp__.py:4: error: Metaclasses not inheriting from 'type' are not supported
```
which I think is wrong (`Any` should be allowed).

Note that the Python 3 equivalent (`class C(metaclass=M): pass`) is not considered an error.

JukkaL commented:
@gvanrossum Your example was fixed by #2914.
